src/code/z_fireobj.c

```diff
@@ -1,27 +1,251 @@
+    fire->sizeInv = 1.0f / init->size;
+    fire->dynamicSizeStep = init->dynamicSizeStep;
+    fire->state = init->state;
+    fire->sizeGrowsCos2 = init->sizeGrowsCos2;
+    fire->unk27 = init->unkA;
+    fire->flags = init->flags;
+    fire->xScale = 0.0f;
+    fire->yScale = 0.0f;
+    fire->dynamicSize = 0.0f;
+    fire->unk26 = Rand_ZeroOne() * 20.0f;
+    fire->ignitionDelay = -1;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_fireobj/FireObj_Draw.s")
+void FireObj_SetState(FireObj* fire, f32 dynamicSizeStep, u8 newState) {
+    fire->state = newState & 0xFF;
+    if (fire->state == FIRE_STATE_3) {
+        fire->yScale = 0.0f;
+        fire->xScale = 0.0f;
+        fire->dynamicSize = 0.0f;
+    } else if (fire->state == FIRE_STATE_2) {
+        fire->xScale = fire->yScale = fire->size;
+        fire->dynamicSize = 1.0f;
+    }
+    fire->dynamicSizeStep = dynamicSizeStep;
+}
 
+void FireObj_SetPosition(FireObj* fire, Vec3f* pos) {
+    Math_Vec3f_Copy(&fire->position, pos);
+}
+
+void FireObj_StepSize(FireObj* fire) {
+    if (fire->state == FIRE_STATE_0) {
+        if (Math_StepToF(&fire->dynamicSize, 1.0f, fire->dynamicSizeStep) != 0) {
+            FireObj_SetState(fire, fire->dynamicSizeStep, FIRE_STATE_2);
+        }
+    } else if ((fire->state == FIRE_STATE_1) && (Math_StepToF(&fire->dynamicSize, 0.0f, fire->dynamicSizeStep) != 0)) {
+        FireObj_SetState(fire, fire->dynamicSizeStep, 3);
+    }
+    if (fire->sizeGrowsCos2 == 1) {
+        if ((fire->state == FIRE_STATE_0) || (fire->state == FIRE_STATE_1)) {
+            fire->xScale = (1.0f - Math_CosS((fire->dynamicSize * fire->dynamicSize * 16384.0f))) * fire->size;
+            fire->yScale = fire->dynamicSize * fire->size;
+        } else {
+            fire->yScale = fire->dynamicSize * fire->size;
+            fire->xScale = fire->dynamicSize * fire->size;
+        }
+        return;
+    } else {
+        fire->yScale = fire->dynamicSize * fire->size;
+    }
+    fire->xScale = fire->dynamicSize * fire->size;
+}
+
+void FireObj_UpdateStateTransitions(GlobalContext* globalCtx, FireObj* fire) {
+    Player* player = GET_PLAYER(globalCtx);
+    WaterBox* waterBox;
+    f32 sp44;
+    s32 sp40 = 0;
+    u8 nextState;
+    Vec3f dist;
+
+    FireObj_StepSize(fire);
+    fire->unk26++;
+    if (fire->ignitionDelay > 0) {
+        fire->ignitionDelay--;
+    } else if (fire->ignitionDelay == 0) {
+        fire->ignitionDelay = -1;
+        if ((fire->state == FIRE_STATE_3) || (fire->state == FIRE_STATE_1)) {
+            nextState = FIRE_STATE_0;
+        } else {
+            nextState = FIRE_STATE_1;
+        }
+        FireObj_SetState(fire, fire->dynamicSizeStep, nextState);
+    }
+    if ((fire->flags & 1) && (fire->state != FIRE_STATE_3) &&
+        (func_800CA1E8(globalCtx, &globalCtx->colCtx, fire->position.x, fire->position.z, &sp44, &waterBox) != 0) &&
+        ((fire->yScale * ((void)0, 6500.0f)) < (sp44 - fire->position.y))) { // Fake but IDK what else
```

Comment:
```suggestion
        ((fire->yScale * (NULL, 6500.0f)) < (sp44 - fire->position.y))) { // Fake but IDK what else
```
could this be a Null pointer?

Reply:
None of the other instances of this trick use that.

Reply:
`NULL` usually expands to `(void*)0` (note the `*` which means it is a pointer), the `void` here just silences an "unused value" warning